### ISSUE TITLE
LLVM to SMT: `argument of incompatible type` for BPF_JEQ on v6.7

### ISSUE THREAD

**Issue by pchaigno** [bpfverif/agni#20]
I'm hitting the following error when trying to run Agni (f4340a2f12d with the patch from https://github.com/bpfverif/agni/pull/19) on Linux v6.7.

```
$ python3 generate_encodings.py --kernver 6.7 --kernbasedir ~/linux --outdir /home/paul/agni/bpf-encodings/6.7
[...]
Getting encoding for BPF_JEQ/home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.ll
/usr/bin/opt -Oz --strip-debug --instnamer --stats -S  /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.ll -o /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ll
/home/paul/agni/llvm-to-smt/llvm-passes/force_functions_early_exit.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ll /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.ll
/usr/bin/opt -Oz --strip-debug --instnamer --stats -S  /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.ll -o /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.ll
/home/paul/agni/llvm-to-smt/llvm-passes/remove_func_calls.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.ll check_cond_jmp_op_wrapper_BPF_JEQ /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.Oz.rfc.ll
/usr/bin/opt -Oz --strip-debug --instnamer --stats -S  /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.ll -o /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ll
/home/paul/agni/llvm-to-smt/llvm-passes/inline_verifier_func.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ll check_cond_jmp_op_wrapper_BPF_JEQ /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.ll
/usr/bin/opt -Oz --strip-debug --instnamer --stats -S  /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.ll -o /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.ll
/home/paul/agni/llvm-to-smt/llvm-passes/promote_memcpy.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.ll check_cond_jmp_op_wrapper_BPF_JEQ /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.ll
/usr/bin/opt -S --instnamer --sroa --adce --bdce --dce --globaldce --deadargelim --unreachableblockelim --lowerswitch --function-attrs --argpromotion --instcombine /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.ll -o /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.ll
/home/paul/agni/llvm-to-smt/llvm-passes/lower_funnel_shifts.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.ll check_cond_jmp_op_wrapper_BPF_JEQ /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ll
/home/paul/agni/llvm-to-smt/llvm-passes/inline_verifier_func.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ll check_cond_jmp_op_wrapper_BPF_JEQ /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.ll
/usr/bin/opt -S --instnamer --sroa --adce --bdce --dce --globaldce --deadargelim --unreachableblockelim --lowerswitch --function-attrs --argpromotion --instcombine /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.ll -o /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.O0.ll
['/usr/bin/llvm-extract', '--func=check_cond_jmp_op_wrapper_BPF_JEQ', '-S', '/home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.O0.ll', '-o', '/home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.O0.ex.ll']
/home/paul/agni/llvm-to-smt/llvm-passes/llvm_to_smt.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.O0.ll check_cond_jmp_op_wrapper_BPF_JEQ 16 BPF_JEQ.smt2
Traceback (most recent call last):
  File "/home/paul/agni/llvm-to-smt/generate_encodings.py", line 685, in <module>
    llvmpassrunner_for_op.run()
  File "/home/paul/agni/llvm-to-smt/run_llvm_passes.py", line 247, in run
    self.run_llvm_to_smt_pass()
  File "/home/paul/agni/llvm-to-smt/run_llvm_passes.py", line 204, in run_llvm_to_smt_pass
    subprocess.run(cmd_llvm_to_smt_op, shell=True, check=True, text=True,
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '/home/paul/agni/llvm-to-smt/llvm-passes/llvm_to_smt.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.O0.ll check_cond_jmp_op_wrapper_BPF_JEQ 16 BPF_JEQ.smt2' returned non-zero exit status 1.
```

And if I rerun the `llvm_to_smt.sh` command:
```
$ /home/paul/agni/llvm-to-smt/llvm-passes/llvm_to_smt.sh /home/paul/agni/bpf-encodings/6.7/BPF_JEQ BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.O0.ll check_cond_jmp_op_wrapper_BPF_JEQ 16 BPF_JEQ.smt2
[...]
-------------------
  %i = ptrtoint %struct.bpf_reg_state* %src_reg to i64
-------------------
[handleCastInst]
[getBitVecSingValType] %struct.bpf_reg_state* %src_reg
[getBitVecSingValType] type: %struct.bpf_reg_state*
opt: /usr/include/llvm-12/llvm/Support/Casting.h:269: typename llvm::cast_retty<X, Y*>::ret_type llvm::cast(Y*) [with X = llvm::IntegerType; Y = const llvm::Type; typename llvm::cast_retty<X, Y*>::ret_type = const llvm::IntegerType*]: Assertion `isa<X>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /usr/bin/opt -load-pass-plugin /home/paul/agni/llvm-to-smt/llvm-passes/LLVMToSMT/build/libLLVMToSMT.so --passes=print<llvm-to-smt> --disable-output /home/paul/agni/bpf-encodings/6.7/BPF_JEQ/BPF_JEQ.Oz.ffee.Oz.rfc.Oz.ivf.Oz.pmc.O0.lfsh.ivf.O0.ll
 #0 0x00007f3eb15732a1 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd732a1)
 #1 0x00007f3eb15713d0 llvm::sys::RunSignalHandlers() (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd713d0)
 #2 0x00007f3eb15739ad (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xd739ad)
 #3 0x00007f3eb0042520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f3eb00969fc __pthread_kill_implementation ./nptl/./nptl/pthread_kill.c:44:76
 #5 0x00007f3eb00969fc __pthread_kill_internal ./nptl/./nptl/pthread_kill.c:78:10
 #6 0x00007f3eb00969fc pthread_kill ./nptl/./nptl/pthread_kill.c:89:10
 #7 0x00007f3eb0042476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x00007f3eb00287f3 abort ./stdlib/./stdlib/abort.c:81:7
 #9 0x00007f3eb002871b _nl_load_domain ./intl/./intl/loadmsgcat.c:1177:9
#10 0x00007f3eb0039e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#11 0x00007f3eadf7c985 llvm::cast_retty<llvm::IntegerType, llvm::Type const*>::ret_type llvm::cast<llvm::IntegerType, llvm::Type const>(llvm::Type const*) /usr/include/llvm-12/llvm/Support/Casting.h:271:72
#12 0x00007f3eadf7c37a llvm::Type::getIntegerBitWidth() const /usr/include/llvm-12/llvm/IR/DerivedTypes.h:97:46
#13 0x00007f3eadf7a78f BitVecHelper::getBitVecSingValType(llvm::Value*) /home/paul/agni/llvm-to-smt/llvm-passes/LLVMToSMT/BitvecHelper.cpp:69:22
#14 0x00007f3eadf83c47 FunctionEncoder::handleCastInst(llvm::CastInst&) /home/paul/agni/llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:217:65
#15 0x00007f3eadf8f9ea FunctionEncoder::populateBBAssertionsMap(llvm::BasicBlock&) /home/paul/agni/llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2014:21
#16 0x00007f3eadf90a6e FunctionEncoder::encodeFunctionBody(llvm::Function&) /home/paul/agni/llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2252:3
#17 0x00007f3eadf92f01 FunctionEncoder::buildSMT() /home/paul/agni/llvm-to-smt/llvm-passes/LLVMToSMT/FunctionEncoder.cpp:2586:54
#18 0x00007f3eadfb15c3 LLVMToSMT::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /home/paul/agni/llvm-to-smt/llvm-passes/LLVMToSMT/LLVMToSMT.cpp:43:11
#19 0x00007f3eadfc2a55 llvm::detail::AnalysisPassModel<llvm::Module, LLVMToSMT, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module>::Invalidator>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /usr/include/llvm-12/llvm/IR/PassManagerInternal.h:315:65
#20 0x00007f3eb16d97e2 llvm::AnalysisManager<llvm::Module>::getResultImpl(llvm::AnalysisKey*, llvm::Module&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xed97e2)
#21 0x00007f3eadfb700a LLVMToSMT::Result& llvm::AnalysisManager<llvm::Module>::getResult<LLVMToSMT>(llvm::Module&) /usr/include/llvm-12/llvm/IR/PassManager.h:789:22
#22 0x00007f3eadfb1b1a LLVMToSMTPrinter::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /home/paul/agni/llvm-to-smt/llvm-passes/LLVMToSMT/LLVMToSMT.cpp:91:50
#23 0x00007f3eadfc2b49 llvm::detail::PassModel<llvm::Module, LLVMToSMTPrinter, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /usr/include/llvm-12/llvm/IR/PassManagerInternal.h:86:3
#24 0x00007f3eb16d67c2 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/lib/x86_64-linux-gnu/libLLVM-12.so.1+0xed67c2)
#25 0x00000000004243f0 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::StringRef>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool) (/usr/bin/opt+0x4243f0)
#26 0x0000000000432874 main (/usr/bin/opt+0x432874)
#27 0x00007f3eb0029d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#28 0x00007f3eb0029e40 call_init ./csu/../csu/libc-start.c:128:20
#29 0x00007f3eb0029e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#30 0x000000000041f9f5 _start (/usr/bin/opt+0x41f9f5)
/home/paul/agni/llvm-to-smt/llvm-passes/llvm_to_smt.sh: line 50: 246562 Aborted
```

**Comment by harishankarv** [comment]
This is because of an incorrect `wrappers.py`.  Your run is likely using `wrapper_7` from `wrappers.py`, which was specifically made for Andrii's patchset, and will not work for 6.7. 

The patchset changed the signature of `is_branch_taken`, from
```
static int is_branch_taken(struct bpf_reg_state *reg, u64 val, u8 opcode, bool is_jmp32)
```
to
```
static int is_branch_taken(struct bpf_reg_state *reg1, struct bpf_reg_state *reg2, u8 opcode, bool is_jmp32)
```
 Due to this breaking change, the llvm emitted a cast instruction, which llvm-to-smt does not handle, and it errored out. 

This should ideally be addressed along with the broader discussion of wrappers at https://github.com/bpfverif/agni/pull/19 and https://github.com/bpfverif/agni/pull/14#issuecomment-1836886144

**Comment by harishankarv** [comment]
If you would like to try it out, here is a patch that has the 6.7 wrapper along with changes to the `generate_encodings.py`. I'm not pushing this because we need to have a better way to do wrappers. 

[6.7_wrapper.patch](https://github.com/bpfverif/agni/files/14475130/6.7_wrapper.patch)

**Comment by pchaigno** [comment]
@harishankarv Did you end up pushing a fix for this? I was able to generate encodings for v6.7 (as shown in my recent PR).

**Comment by harishankarv** [comment]
@pchaigno yes that's right. This was fixed by [5c3fc41](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html). I probably should have split it into two commits.
